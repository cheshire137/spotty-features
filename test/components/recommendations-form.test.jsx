import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import RecommendationsForm from '../../src/components/recommendations-form.jsx'

import ArtistSearchResponse from '../fixtures/spotify/artist-search'
import TrackSearchResponse from '../fixtures/spotify/track-search'

const track = TrackSearchResponse.tracks.items[0]
const trackSeed = {
  id: track.id,
  name: track.name,
  url: track.external_urls.spotify,
  artists: track.artists.map(artist => artist.name),
  album: track.album.name,
  image: track.album.images[2].url,
  albumUrl: track.album.external_urls.spotify
}
const trackFeatures = {
  danceability: 0.468,
  energy: 0.515,
  speechiness: 0.0665,
  acousticness: 0.145,
  instrumentalness: 0,
  liveness: 0.132,
  valence: 0.186
}

const artist = ArtistSearchResponse.artists.items[0]
const artistSeed = {
  id: artist.id,
  name: artist.name,
  url: artist.external_urls.spotify,
  image: artist.images[3].url
}
const artistFeatures = {
  danceability: 0.5,
  energy: 0.5,
  speechiness: 0,
  acousticness: 0.5,
  instrumentalness: 0.5,
  liveness: 0,
  valence: 0.5
}

function props(opts) {
  return {
    changeSeed: opts.changeSeed,
    numRecommendations: 25,
    seed: opts.seed,
    seedType: opts.seedType,
    token: '123abc',
    onNumRecommendationsChange: opts.onNumRecommendationsChange,
    onRecommendations: opts.onRecommendations,
    features: opts.features
  }
}

describe('RecommendationsForm', () => {
  let wasSeedChanged = false
  let numRecommendations = null
  let recommendations = null

  const changeSeed = () => {
    wasSeedChanged = true
  }

  const onNumRecommendationsChange = newNum => {
    numRecommendations = newNum
  }

  const onRecommendations = newRecs => {
    recommendations = newRecs
  }

  const baseOpts = {
    changeSeed, onNumRecommendationsChange, onRecommendations
  }

  function getTrackComponent() {
    const opts = Object.assign(baseOpts, {
      seed: trackSeed, seedType: 'track', features: trackFeatures
    })
    return <RecommendationsForm {...props(opts)} />
  }

  function getArtistComponent() {
    const opts = Object.assign(baseOpts, {
      seed: artistSeed, seedType: 'artist', features: artistFeatures
    })
    return <RecommendationsForm {...props(opts)} />
  }

  afterEach(() => {
    wasSeedChanged = false
    numRecommendations = null
    recommendations = null
  })

  test('matches snapshot for track seed', () => {
    const component = getTrackComponent()
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('matches snapshot for artist seed', () => {
    const component = getArtistComponent()
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('can change how many recommendations to get', () => {
    const component = getTrackComponent()
    const select = shallow(component).find('#num-recommendations')
    select.simulate('change', { target: { value: 50 } })
    expect(numRecommendations).toBe(50)
  })

  test('can change feature slider', () => {
    const wrapper = shallow(getArtistComponent())
    const slider = wrapper.find('#liveness')
    slider.simulate('change', { target: { value: 0.3 } })
    expect(wrapper.find('#liveness').props().value).toBeCloseTo(0.3, 5)
  })
})
