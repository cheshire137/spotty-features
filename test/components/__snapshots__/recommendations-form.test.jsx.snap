exports[`RecommendationsForm matches snapshot for artist seed 1`] = `
<div
  className="content">
  <p>
    <button
      className="button is-link"
      onClick={[Function]}
      type="button">
      ← Change seed
    </button>
  </p>
  <form
    onSubmit={[Function]}>
    <h3
      className="refine-title title is-3">
      Step 2: Refine your results
    </h3>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="acousticness"
          style={
            Object {
              "color": "#FF7EAB",
            }
          }>
          Acoustic
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="acousticness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.5} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          50
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="danceability"
          style={
            Object {
              "color": "#0ABD9C",
            }
          }>
          Danceable
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="danceability"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.5} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          50
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="energy"
          style={
            Object {
              "color": "#e3b951",
            }
          }>
          Energetic
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="energy"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.5} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          50
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="instrumentalness"
          style={
            Object {
              "color": "#494949",
            }
          }>
          Instrumental
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="instrumentalness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.5} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          50
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="liveness"
          style={
            Object {
              "color": "#AF33C8",
            }
          }>
          Live
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="liveness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          0
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="valence"
          style={
            Object {
              "color": "#90B207",
            }
          }>
          Positivity
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="valence"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.5} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          50
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="speechiness"
          style={
            Object {
              "color": "#f09945",
            }
          }>
          Speechy
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="speechiness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          0
          %
        </span>
      </div>
    </div>
    <div
      className="columns form-controls">
      <div
        className="column">
        <div
          className="control is-large is-horizontal">
          <div
            className="control-label">
            <label
              className="label nowrap"
              htmlFor="num-recommendations">
              How many songs to find:
            </label>
          </div>
          <div
            className="control">
            <span
              className="select">
              <select
                id="num-recommendations"
                onChange={[Function]}
                value={25}>
                <option
                  value="10">
                  10
                </option>
                <option
                  value="25">
                  25
                </option>
                <option
                  value="50">
                  50
                </option>
                <option
                  value="75">
                  75
                </option>
                <option
                  value="100">
                  100
                </option>
              </select>
            </span>
          </div>
        </div>
      </div>
      <div
        className="column">
        <div
          className="control">
          <button
            className="button is-primary is-right is-large"
            type="submit">
            Find Songs
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;

exports[`RecommendationsForm matches snapshot for track seed 1`] = `
<div
  className="content">
  <p>
    <button
      className="button is-link"
      onClick={[Function]}
      type="button">
      ← Change seed
    </button>
  </p>
  <form
    onSubmit={[Function]}>
    <h3
      className="refine-title title is-3">
      Step 2: Refine your results
    </h3>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="acousticness"
          style={
            Object {
              "color": "#FF7EAB",
            }
          }>
          Acoustic
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="acousticness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.145} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          14
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="danceability"
          style={
            Object {
              "color": "#0ABD9C",
            }
          }>
          Danceable
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="danceability"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.468} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          47
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="energy"
          style={
            Object {
              "color": "#e3b951",
            }
          }>
          Energetic
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="energy"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.515} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          52
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="instrumentalness"
          style={
            Object {
              "color": "#494949",
            }
          }>
          Instrumental
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="instrumentalness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          0
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="liveness"
          style={
            Object {
              "color": "#AF33C8",
            }
          }>
          Live
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="liveness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.132} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          13
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="valence"
          style={
            Object {
              "color": "#90B207",
            }
          }>
          Positivity
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="valence"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.186} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          19
          %
        </span>
      </div>
    </div>
    <div
      className="control is-horizontal">
      <div
        className="control-label">
        <label
          className="label"
          htmlFor="speechiness"
          style={
            Object {
              "color": "#f09945",
            }
          }>
          Speechy
        </label>
      </div>
      <div
        className="control">
        <span
          className="feature-range-min">
          0%
        </span>
        <input
          className="slider"
          id="speechiness"
          max="1"
          min="0"
          onChange={[Function]}
          step="0.05"
          type="range"
          value={0.0665} />
        <span
          className="feature-range-max">
          100%
        </span>
        <span
          className="feature-percentage">
          7
          %
        </span>
      </div>
    </div>
    <div
      className="columns form-controls">
      <div
        className="column">
        <div
          className="control is-large is-horizontal">
          <div
            className="control-label">
            <label
              className="label nowrap"
              htmlFor="num-recommendations">
              How many songs to find:
            </label>
          </div>
          <div
            className="control">
            <span
              className="select">
              <select
                id="num-recommendations"
                onChange={[Function]}
                value={25}>
                <option
                  value="10">
                  10
                </option>
                <option
                  value="25">
                  25
                </option>
                <option
                  value="50">
                  50
                </option>
                <option
                  value="75">
                  75
                </option>
                <option
                  value="100">
                  100
                </option>
              </select>
            </span>
          </div>
        </div>
      </div>
      <div
        className="column">
        <div
          className="control">
          <button
            className="button is-primary is-right is-large"
            type="submit">
            Find Songs
          </button>
        </div>
      </div>
    </div>
  </form>
</div>
`;
